editor/src/core/property-controls/property-controls-local.ts

```diff
@@ -174,49 +174,91 @@ export function parseComponentToRegister(value: unknown): ParseResult<ComponentT
+    },
+    (components: { [componentName: string]: ComponentToRegister }) => {
+      return prepareComponentDescriptorsForRegistering(workers, moduleNameOrPath, components)
+    },
+    parsedComponents,
+  )
+}
+
+type PartiallyAppliedParseAndPrepareComponents = (
+  unparsedComponents: unknown,
+) => Either<string, PreparedComponentDescriptorsForRegistering>
+
+const partiallyParseAndPrepareComponents = (
+  workers: UtopiaTsWorkers,
+  moduleNameOrPath: string,
+): PartiallyAppliedParseAndPrepareComponents => {
+  return (unparsedComponents: unknown) =>
+    parseAndPrepareComponents(workers, moduleNameOrPath, unparsedComponents)
+}
+
 export function createRegisterModuleFunction(
   workers: UtopiaTsWorkers | null,
 ): typeof registerModuleAPI {
+  let cachedParseAndPrepareComponentsMap: Map<
+    string,
+    PartiallyAppliedParseAndPrepareComponents
+  > = new Map()
+
   // create a function with a signature that matches utopia-api/registerModule
   return function registerModule(
     unparsedModuleName: string,
     unparsedComponents: { [componentName: string]: ComponentToRegister },
   ): void {
     const parsedModuleName = parseString(unparsedModuleName)
-    const parsedComponents = parseComponents(unparsedComponents)
-
-    const parsedParams = applicative2Either(
-      (moduleName, components) => {
-        return { moduleName: moduleName, components: components }
-      },
-      parsedModuleName,
-      parsedComponents,
-    )
 
     if (isLeft(parsedModuleName)) {
       const errorDetails = getParseErrorDetails(parsedModuleName.value)
       throw new Error(`registerModule first param (moduleName): ${errorDetails.description}`)
-    }
-
-    if (isLeft(parsedComponents)) {
-      const errorDetails = getParseErrorDetails(parsedComponents.value)
-      throw new Error(
-        `registerModule second param (components): ${errorDetails.description} [${errorDetails.path}]`,
-      )
-    }
-
-    forEachRight(parsedParams, ({ moduleName, components }) => {
+    } else {
       if (workers != null) {
-        // Fires off asynchronously.
-        registerModuleInternal(workers, moduleName, components)
+        const moduleName = parsedModuleName.value
+        let parseAndPrepareComponentsFn = cachedParseAndPrepareComponentsMap.get(moduleName)
+        if (parseAndPrepareComponentsFn == null) {
+          // Create a memoized function for the handling of component descriptors for the specified module name
+          parseAndPrepareComponentsFn = memoize(
+            partiallyParseAndPrepareComponents(workers, moduleName),
+            {
+              equals: fastDeepEqual,
+              maxSize: 5,
+            },
+          )
+          cachedParseAndPrepareComponentsMap.set(moduleName, parseAndPrepareComponentsFn)
+        }
+
+        const parsedPreparedDescriptors = parseAndPrepareComponentsFn(unparsedComponents)
```

Comment:
Looks like a candidate for a `foldEither`, but that's very minor.

Reply:
Ah you're right, both this and the surrounding `if` statement would work better as a `foldEither`. Will update them :+1: